engine/src/main/java/org/pentaho/di/trans/steps/http/HTTP.java

```
@@ -0,0 +1,413 @@
+              body = "";
+            }
+            break;
+        }
+
+        Header[] headers = searchForHeaders( httpResponse );
+
+        JSONObject json = new JSONObject();
+        for ( Header header : headers ) {
+          Object previousValue = json.get( header.getName() );
+          if ( previousValue == null ) {
+            json.put( header.getName(), header.getValue() );
+          } else if ( previousValue instanceof List ) {
+            List<String> list = (List<String>) previousValue;
+            list.add( header.getValue() );
+          } else {
+            ArrayList<String> list = new ArrayList<String>();
+            list.add( (String) previousValue );
+            list.add( header.getValue() );
+            json.put( header.getName(), list );
+          }
+        }
+        String headerString = json.toJSONString();
+
+        int returnFieldsOffset = rowMeta.size();
+        if ( !Utils.isEmpty( meta.getFieldName() ) ) {
+          newRow = RowDataUtil.addValueData( newRow, returnFieldsOffset, body );
+          returnFieldsOffset++;
+        }
+
+        if ( !Utils.isEmpty( meta.getResultCodeFieldName() ) ) {
+          newRow = RowDataUtil.addValueData( newRow, returnFieldsOffset, new Long( statusCode ) );
+          returnFieldsOffset++;
+        }
+        if ( !Utils.isEmpty( meta.getResponseTimeFieldName() ) ) {
+          newRow = RowDataUtil.addValueData( newRow, returnFieldsOffset, new Long( responseTime ) );
+          returnFieldsOffset++;
+        }
+        if ( !Utils.isEmpty( meta.getResponseHeaderFieldName() ) ) {
+          newRow = RowDataUtil.addValueData( newRow, returnFieldsOffset, headerString );
+        }
+
+      } finally {
+        if ( httpResponse != null ) {
+          httpResponse.close();
+        }
+        // Release current connection to the connection pool once you are done
+        method.releaseConnection();
+      }
+      return newRow;
+    } catch ( UnknownHostException uhe ) {
```

Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)